lib/partitioninfo.ts

```diff
@@ -0,0 +1,312 @@
+	const partitions = getPartitionsFromMBRBuf(mbrBuf);
+	if (partitions.length === 1 && partitions[0].type === GPT_PROTECTIVE_MBR) {
+		const gptBuf = await readFromDisk(disk, MBR_SIZE, GPT_SIZE);
+		const gpt = GPT.parse(gptBuf);
+		return {
+			type: 'gpt',
+			partitions: gpt.partitions.map((partition, index) =>
+				gptPartitionDict(gpt, partition, index + 1),
+			),
+		};
+	} else {
+		const mbrPartitions: MBRPartition[] = [];
+		for (let index = 0; index < partitions.length; index++) {
+			const p = partitions[index];
+			if (p.extended) {
+				extended = p;
+				if (includeExtended) {
+					mbrPartitions.push(mbrPartitionDict(p, offset, index + 1));
+				}
+			} else {
+				mbrPartitions.push(mbrPartitionDict(p, offset, index + 1));
+			}
+		}
+		if (extended != null && getLogical) {
+			const logicalPartitions = await getLogicalPartitions(
+				disk,
+				MBR_FIRST_LOGICAL_PARTITION,
+				extended.byteOffset(),
+			);
+			mbrPartitions.push(...logicalPartitions);
+		}
+		return { type: 'mbr', partitions: mbrPartitions };
+	}
+}
+
+class PartitionNotFound extends TypedError {
+	constructor(partitionNumber: number) {
+		super(`Partition not found: ${partitionNumber}.`);
+	}
+}
+
+async function getPartition(
+	disk: Disk,
+	partitionNumber: number,
+): Promise<MBRPartition | GPTPartition> {
+	if (partitionNumber < 1) {
+		throw new Error('The partition number must be at least 1.');
+	}
+	const info = await getDiskPartitions(disk, {
+		includeExtended: true,
+		offset: 0,
+		getLogical: false,
+	});
+	if (info.type === 'gpt') {
+		if (info.partitions.length < partitionNumber) {
+			throw new PartitionNotFound(partitionNumber);
+		} else {
+			return info.partitions[partitionNumber - 1];
+		}
+	}
+	if (partitionNumber <= MBR_LAST_PRIMARY_PARTITION) {
+		if (partitionNumber <= info.partitions.length) {
+			return info.partitions[partitionNumber - 1];
+		} else {
+			throw new PartitionNotFound(partitionNumber);
+		}
+	}
+	const extended = info.partitions.find(p => MBR.Partition.isExtended(p.type));
+	if (!extended) {
+		throw new PartitionNotFound(partitionNumber);
+	} else {
+		const logicalPartitionPosition =
+			partitionNumber - MBR_FIRST_LOGICAL_PARTITION;
+		const logicalPartitions = await getLogicalPartitions(
+			disk,
+			MBR_FIRST_LOGICAL_PARTITION,
+			extended.offset,
+			extended.offset,
+			logicalPartitionPosition + 1,
+		);
+		if (logicalPartitionPosition < logicalPartitions.length) {
+			return logicalPartitions[logicalPartitionPosition];
+		} else {
+			throw new PartitionNotFound(partitionNumber);
+		}
+	}
+}
+
+function isString(x: any): x is string {
+	return typeof x === 'string';
+}
+
+async function callWithDisk<ParameterType, ReturnType>(
+	fn: (disk: Disk, arg: ParameterType) => Promise<ReturnType>,
+	pathOrDisk: string | Disk,
+	arg: ParameterType,
+): Promise<ReturnType> {
+	if (isString(pathOrDisk)) {
+		return await using(openFile(pathOrDisk, 'r'), async fd => {
```

Comment:
NIT: I would personally import this as `Bluebird` still, I'd say `Bluebird.using` is a lot clearer than just `using` but I'm not too concerned.